Add key decode options for atom decoding

eproxus opened the pull request:
This allows one to use the option `keys` set to either `to_existing_atom` or `to_atom` to decode keys as atoms.

10s benchmark comparing the new options to `thoas` (new code without the `key` option) and `thoas_orig` (a copy of the old unmodified code).
<img width="1217" alt="Screenshot 2023-03-24 at 19 23 31" src="https://user-images.githubusercontent.com/112878/227608849-3e13147d-4c7a-4962-baff-5ac793b20956.png">

(full benchmark [output.7z.zip](https://github.com/lpil/thoas/files/11065824/output.7z.zip), _note: 7z archive inside a zip archive, to avoid GitHubs 25 Mb limit_)

Implementation notes:
- I tried putting `StringDecode` and `KeyDecode` in a tuple `{StringDecode, KeyDecode}` and pattern match it out only where needed. That was way slower (~2x) for some reason so now it is just a new argument to all functions.
- It seems the performance is on par with the original implementation, with some minor differences. Sometimes converting keys is faster than leaving them as binaries, sometimes slower.

Things to check before this PR should be accepted:
- [ ] Is the implementation done in the right place?
- [ ] Is the code formatting okay?
- [ ] There is a new error when an invalid value for `keys` is used. Is this error format acceptable?
- [ ] Someone else should re-run the full benchmark suite to verify that the changes indeed did not affect the original performance too much (I copied the `thoas_decode` module to `thoas_decode_orig` and made copies of the `decode/1+2` functions in `thoas` itself to add it to the benchmark).
- [ ] Do we want to add benchmark entries for the other libraries that support these options? (would be interesting IMO)
- [ ] Proof-read the new documentation

Suggestions welcome!

eproxus commented:
Friendly reminder hoping this can get reviewed soon 😊

eproxus commented:
@lpil What is the status of this project? Is it still active?

lpil commented on src/thoas_decode.erl:
```diff
@@ -1073,23 +1077,45 @@ string_decode_function(Options) ->
+        error -> fun identity/1;
+        {ok, to_atom} -> fun to_atom/1;
+        {ok, to_existing_atom} -> fun to_existing_atom/1;
+        {ok, Invalid} -> error({invalid_option, #{keys => Invalid}})
```
Let's either add this error case to `string_decode_function` or remove this case so it errors from the atom not being matched

eproxus replied:
That makes sense. While refactoring, I had the thought we could use the same logic to decode both keys and strings (with separate options, of course). Something like this:

```erlang
decode_function(Option, Options) ->
    case maps:find(Option, Options) of
        error -> fun reference/1;
        {ok, reference} -> fun to_atom/1;
        {ok, copy} -> fun binary:copy/1;
        {ok, atom} -> fun to_atom/1;
        {ok, existing_atom} -> fun to_existing_atom/1;
        {ok, Invalid} -> error({invalid_option, #{Option => Invalid}})
    end.
```

That way, you could decode strings to existing atoms too if you wanted. What do you think?

eproxus replied:
I took the liberty of doing that change, so you can tell me what you think. If you'd rather have the old separate logic, let me know and I'll change it back 🙂

lpil replied:
I don't think we should decode string values to atoms. Just reference or copy as today 👍

eproxus replied:
I think it could be useful for some applications where you already have a subset of the data in code as atoms. E.g.:
```erlang
1> thoas:decode(Pokedex, #{keys => existing_atom, strings => existing_atom}).
{ok,#{pokemon =>
          [#{id => 1,name => <<"Bulbasaur">>,
             type => ['Grass','Poison'], % Atoms
             weight => <<"6.9 kg">>,
             weaknesses => ['Fire','Ice','Flying','Psychic'],  % Atoms
             spawn_time => <<"20:00">>,spawn_chance => 0.69,
             num => <<"001">>,
             next_evolution =>
                 [#{name => <<"Ivysaur">>,num => <<"002">>},
                  #{name => <<"Venusaur">>,num => <<"003">>}],
             multipliers => [1.58],
             img =>
                 <<"http://www.serebii.net/pokemongo/pokemon/001.png">>,
             height => <<"0.71 m">>,egg => <<"2 km">>,candy_count => 25,
             candy => <<"Bulbasaur Candy">>,avg_spawns => 69},
           ...]}}
```

eproxus replied:
If you are still not convinced, I’ll remove it!

By the way, do you think the option `copy` has merit for keys?

lpil replied:
Thank you.

Ah yes, good point. Copying keys is probably useful for some people

eproxus replied:
Should I leave it as is, or should we revert to a version where the errors are the same but with the old string decode logic and `copy` as an option for keys?

lpil replied:
Two functions please, so that values cannot be used for values.

lpil reviewed (commented):
Hello! Thank you for this. Sorry about not reviewing this, it got lost in my frightfully busy inbox.

This all seems good and sensible to me. I'm away from home so I'm unable to run the benchmarks presently, but afterwards I see no reason not to release this.

williamthome commented:
@eproxus could you please take a look at why it is failing? 
This and #25 are a good fit for a project I'm working on.

eproxus committed:
Add key decode options for atom decoding

This allows one to use the option `keys` set to either
`to_existing_atom` or `to_atom` to decode keys as atoms.

eproxus commented:
@lpil I've updated the PR according to your comments.

I fixed the test failure by adding the option `%{keys: :copy}`. I'm not sure if this what was the test was intended to test, but it passes now at least. Let me know if I should correct anything.

(cc @williamthome)

lpil commented on src/thoas.erl:
```diff
@@ -11,7 +11,8 @@
 -type decode_options() :: #{
-    strings => reference | copy
+    strings => reference | copy,
+    keys => to_existing_atom | to_atom
```
I think this is missing `reference | copy`?

eproxus replied:
Great catch! You are right.

lpil reviewed (approved):
Thank you
